File: frontend/src/types/complaint.ts
export type AnalysisStatus = "pending" | "processing" | "completed" | "failed";

export type SentimentLabel = "positive" | "neutral" | "negative" | "mixed";

export interface ComplaintImageAnalysis {
  primary_issue: string;
  secondary_issue?: string | null;
  description: string;
  severity: string;
  confidence_score: number;
  visible_damage: string;
  estimated_area_affected: string;
  safety_risk: string;
  suggested_department: string;
  suggested_immediate_action: string;
  suggested_long_term_action: string;
  possible_public_impact: string;
  duplicate_indicators: string[];
  detected_objects: string[];
  environmental_risk: string;
  road_safety_risk: string;
  human_presence: boolean;
  vehicles_present: boolean;
  requires_urgent_attention: boolean;
  reasoning: string;
  processed_at: string;
  model_version: string;
  prompt_version: string;
}

export interface ComplaintAIAnalysis {
  category: string;
  sub_category: string;
  responsible_department: string;
  urgency: string;
  severity: string;
  priority_level: string;
  summary: string;
  detailed_explanation: string;
  keywords: string[];
  affected_infrastructure: string;
  affected_citizens_estimate?: string | null;
  government_scheme?: string | null;
  suggested_immediate_action: string;
  suggested_long_term_action: string;
  required_department: string;
  required_team: string;
  confidence_score: number;
  reasoning: string;
  duplicate_possibility: number;
  tags: string[];
  language_detected: string;
  translated_english?: string | null;
  voice_transcript?: string | null;
  sentiment: SentimentLabel;
  urgency_score: number;
  language: string;
  processed_at: string;
  model_version: string;
  prompt_version: string;
}

export type ComplaintStatus =
  | "pending"
  | "submitted"
  | "under_review"
  | "clustered"
  | "in_progress"
  | "resolved"
  | "rejected"
  | "closed";

export type ComplaintPriority = "low" | "medium" | "high" | "critical";

export interface GeoLocation {
  latitude: number;
  longitude: number;
  address?: string | null;
}

export interface DocumentMetadata {
  created_at: string;
  updated_at: string;
  created_by: string;
  updated_by: string;
  version: number;
}

export interface Category {
  id: string;
  name: string;
  slug: string;
  description?: string | null;
  icon?: string | null;
  display_order: number;
  is_active: boolean;
  metadata: DocumentMetadata;
}

export interface ComplaintMediaPayload {
  data: string;
  mime_type: string;
  file_name?: string;
}

export interface ComplaintSubmitPayload {
  title: string;
  description: string;
  category_id: string;
  latitude: number;
  longitude: number;
  address: string;
  landmark?: string;
  image?: ComplaintMediaPayload;
  audio?: ComplaintMediaPayload;
  audio_duration_seconds?: number;
  contact_name?: string;
  mobile_number: string;
  is_anonymous: boolean;
}

export interface ComplaintListItem {
  id: string;
  title: string;
  description: string;
  category_id: string;
  category_name: string;
  category_slug: string;
  status: ComplaintStatus;
  priority: ComplaintPriority;
  address?: string | null;
  landmark?: string | null;
  is_anonymous: boolean;
  citizen_name?: string | null;
  citizen_phone?: string | null;
  has_image: boolean;
  has_audio: boolean;
  submitted_at: string;
  constituency: string;
  analysis_status?: AnalysisStatus | null;
  has_ai_analysis?: boolean;
  cluster_id?: string | null;
  is_duplicate?: boolean;
  duplicate_score?: number | null;
}

export interface ComplaintDetail {
  id: string;
  title: string;
  description: string;
  category_id: string;
  category_name: string;
  category_slug: string;
  status: ComplaintStatus;
  priority: ComplaintPriority;
  location?: GeoLocation | null;
  landmark?: string | null;
  image_base64?: string | null;
  image_mime_type?: string | null;
  audio_base64?: string | null;
  audio_mime_type?: string | null;
  audio_duration_seconds?: number | null;
  is_anonymous: boolean;
  citizen_name?: string | null;
  citizen_phone?: string | null;
  constituency: string;
  district: string;
  state: string;
  village_name: string;
  submitted_at: string;
  resolved_at?: string | null;
  metadata: DocumentMetadata;
  ai_analysis?: ComplaintAIAnalysis | null;
  analysis_status: AnalysisStatus;
  analysis_started_at?: string | null;
  analysis_completed_at?: string | null;
  analysis_model_name?: string | null;
  analysis_processing_time_ms?: number | null;
  analysis_prompt_version?: string | null;
  analysis_error_message?: string | null;
  analysis_retry_count: number;
  image_analysis?: ComplaintImageAnalysis | null;
  image_analysis_status?: AnalysisStatus | null;
  vision_model?: string | null;
  vision_processing_time_ms?: number | null;
  vision_completed_at?: string | null;
  vision_started_at?: string | null;
  vision_prompt_version?: string | null;
  vision_error_message?: string | null;
  vision_retry_count: number;
  cluster_id?: string | null;
  is_duplicate?: boolean;
  duplicate_score?: number | null;
  duplicate_reason?: string | null;
  duplicate_confidence?: number | null;
  matched_complaint_id?: string | null;
  matched_cluster_id?: string | null;
}

export interface ComplaintListResponse {
  success: boolean;
  items: ComplaintListItem[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface ComplaintCreateResponse {
  success: boolean;
  message: string;
  complaint: ComplaintDetail;
}

export interface CategoryListResponse {
  success: boolean;
  items: Category[];
}

File: backend/app/models/domain/complaint.py
"""Complaint domain models."""

from datetime import datetime

from pydantic import BaseModel, EmailStr, Field

from app.models.enums.common import (
    AnalysisStatus,
    ComplaintCategory,
    ComplaintPriority,
    ComplaintStatus,
    SentimentLabel,
)
from app.models.schemas.common import Attachment, DocumentMetadata, DocumentMetadataCreate, GeoLocation


class ComplaintImageAnalysis(BaseModel):
    """Gemini Vision image intelligence attached to a complaint."""

    primary_issue: str = Field(max_length=128)
    secondary_issue: str | None = Field(default=None, max_length=128)
    description: str = Field(max_length=3000)
    severity: str = Field(default="Medium", max_length=32)
    confidence_score: float = Field(default=0.0, ge=0, le=1)
    visible_damage: str = Field(default="", max_length=2000)
    estimated_area_affected: str = Field(default="", max_length=512)
    safety_risk: str = Field(default="", max_length=2000)
    suggested_department: str = Field(default="", max_length=256)
    suggested_immediate_action: str = Field(default="", max_length=2000)
    suggested_long_term_action: str = Field(default="", max_length=2000)
    possible_public_impact: str = Field(default="", max_length=2000)
    duplicate_indicators: list[str] = Field(default_factory=list, max_length=20)
    detected_objects: list[str] = Field(default_factory=list, max_length=50)
    environmental_risk: str = Field(default="", max_length=2000)
    road_safety_risk: str = Field(default="", max_length=2000)
    human_presence: bool = False
    vehicles_present: bool = False
    requires_urgent_attention: bool = False
    reasoning: str = Field(default="", max_length=3000)
    processed_at: datetime
    model_version: str = Field(default="gemini-2.5-flash", max_length=64)
    prompt_version: str = Field(default="1.0.0", max_length=16)


class ComplaintAIAnalysis(BaseModel):
    """AI-generated civic intelligence attached to a complaint."""

    category: str = Field(default="", max_length=128)
    sub_category: str = Field(default="", max_length=128)
    responsible_department: str = Field(default="", max_length=256)
    urgency: str = Field(default="medium", max_length=32)
    severity: str = Field(default="medium", max_length=32)
    priority_level: str = Field(default="medium", max_length=32)
    summary: str = Field(max_length=2000)
    detailed_explanation: str = Field(default="", max_length=5000)
    keywords: list[str] = Field(default_factory=list, max_length=50)
    affected_infrastructure: str = Field(default="", max_length=1000)
    affected_citizens_estimate: str | None = Field(default=None, max_length=256)
    government_scheme: str | None = Field(default=None, max_length=512)
    suggested_immediate_action: str = Field(default="", max_length=2000)
    suggested_long_term_action: str = Field(default="", max_length=2000)
    required_department: str = Field(default="", max_length=256)
    required_team: str = Field(default="", max_length=256)
    confidence_score: float = Field(default=0.0, ge=0, le=1)
    reasoning: str = Field(default="", max_length=3000)
    duplicate_possibility: float = Field(default=0.0, ge=0, le=1)
    tags: list[str] = Field(default_factory=list, max_length=20)
    language_detected: str = Field(default="en", max_length=32)
    translated_english: str | None = Field(default=None, max_length=5000)
    voice_transcript: str | None = Field(default=None, max_length=5000)
    sentiment: SentimentLabel = SentimentLabel.NEUTRAL
    urgency_score: float = Field(default=0.5, ge=0, le=1)
    language: str = Field(default="en", max_length=16)
    processed_at: datetime
    model_version: str = Field(default="gemini-2.5-flash", max_length=64)
    prompt_version: str = Field(default="1.0.0", max_length=16)


class ComplaintBase(BaseModel):
    """Shared complaint fields."""

    title: str = Field(min_length=5, max_length=256)
    description: str = Field(min_length=10, max_length=5000)
    category: ComplaintCategory
    priority: ComplaintPriority = ComplaintPriority.MEDIUM
    status: ComplaintStatus = ComplaintStatus.SUBMITTED
    village_id: str = Field(min_length=1, max_length=128)
    village_ref: str = Field(description="Firestore path: villages/{id}")
    village_name: str = Field(max_length=128)
    constituency: str = Field(max_length=128)
    district: str = Field(max_length=128)
    state: str = Field(max_length=64)
    citizen_name: str | None = Field(default=None, max_length=128)
    citizen_phone: str | None = Field(default=None, pattern=r"^\+?[0-9]{10,15}$")
    citizen_email: EmailStr | None = None
    location: GeoLocation | None = None
    landmark: str | None = Field(default=None, max_length=256)
    category_id: str | None = Field(default=None, max_length=128)
    category_name: str | None = Field(default=None, max_length=128)
    category_slug: str | None = Field(default=None, max_length=64)
    image_base64: str | None = None
    image_mime_type: str | None = Field(default=None, max_length=128)
    audio_base64: str | None = None
    audio_mime_type: str | None = Field(default=None, max_length=128)
    audio_duration_seconds: float | None = Field(default=None, ge=0)
    is_anonymous: bool = False
    attachments: list[Attachment] = Field(default_factory=list, max_length=10)
    cluster_id: str | None = Field(default=None, max_length=128)
    cluster_ref: str | None = Field(default=None, description="Firestore path: clusters/{id}")
    ai_analysis: ComplaintAIAnalysis | None = None
    analysis_status: AnalysisStatus = AnalysisStatus.PENDING
    analysis_started_at: datetime | None = None
    analysis_completed_at: datetime | None = None
    analysis_model_name: str | None = Field(default=None, max_length=64)
    analysis_processing_time_ms: int | None = Field(default=None, ge=0)
    analysis_prompt_version: str | None = Field(default=None, max_length=16)
    analysis_error_message: str | None = Field(default=None, max_length=2000)
    analysis_retry_count: int = Field(default=0, ge=0)
    image_analysis: ComplaintImageAnalysis | None = None
    image_analysis_status: AnalysisStatus | None = None
    vision_model: str | None = Field(default=None, max_length=64)
    vision_processing_time_ms: int | None = Field(default=None, ge=0)
    vision_completed_at: datetime | None = None
    vision_started_at: datetime | None = None
    vision_prompt_version: str | None = Field(default=None, max_length=16)
    vision_error_message: str | None = Field(default=None, max_length=2000)
    vision_retry_count: int = Field(default=0, ge=0)
    is_duplicate: bool = False
    duplicate_score: float | None = Field(default=None, ge=0, le=100)
    duplicate_reason: str | None = Field(default=None, max_length=2000)
    duplicate_confidence: float | None = Field(default=None, ge=0, le=1)
    matched_complaint_id: str | None = Field(default=None, max_length=128)
    matched_cluster_id: str | None = Field(default=None, max_length=128)
    submitted_at: datetime
    resolved_at: datetime | None = None


class ComplaintCreate(BaseModel):
    """Fields required to create a complaint."""

    title: str = Field(min_length=5, max_length=256)
    description: str = Field(min_length=10, max_length=5000)
    category: ComplaintCategory
    priority: ComplaintPriority = ComplaintPriority.MEDIUM
    village_id: str = Field(min_length=1, max_length=128)
    village_name: str = Field(max_length=128)
    constituency: str = Field(max_length=128)
    district: str = Field(max_length=128)
    state: str = Field(max_length=64)
    citizen_name: str | None = Field(default=None, max_length=128)
    citizen_phone: str | None = Field(default=None, pattern=r"^\+?[0-9]{10,15}$")
    citizen_email: EmailStr | None = None
    location: GeoLocation | None = None
    landmark: str | None = Field(default=None, max_length=256)
    category_id: str | None = Field(default=None, max_length=128)
    category_name: str | None = Field(default=None, max_length=128)
    category_slug: str | None = Field(default=None, max_length=64)
    image_base64: str | None = None
    image_mime_type: str | None = Field(default=None, max_length=128)
    audio_base64: str | None = None
    audio_mime_type: str | None = Field(default=None, max_length=128)
    audio_duration_seconds: float | None = Field(default=None, ge=0)
    is_anonymous: bool = False
    attachments: list[Attachment] = Field(default_factory=list, max_length=10)
    status: ComplaintStatus = ComplaintStatus.PENDING
    metadata: DocumentMetadataCreate = Field(default_factory=DocumentMetadataCreate)


class ComplaintUpdate(BaseModel):
    """Fields allowed on complaint update."""

    title: str | None = Field(default=None, min_length=5, max_length=256)
    description: str | None = Field(default=None, min_length=10, max_length=5000)
    category: ComplaintCategory | None = None
    category_id: str | None = Field(default=None, max_length=128)
    category_name: str | None = Field(default=None, max_length=128)
    category_slug: str | None = Field(default=None, max_length=64)
    priority: ComplaintPriority | None = None
    status: ComplaintStatus | None = None
    location: GeoLocation | None = None
    cluster_id: str | None = Field(default=None, max_length=128)
    cluster_ref: str | None = None
    ai_analysis: ComplaintAIAnalysis | None = None
    analysis_status: AnalysisStatus | None = None
    analysis_started_at: datetime | None = None
    analysis_completed_at: datetime | None = None
    analysis_model_name: str | None = Field(default=None, max_length=64)
    analysis_processing_time_ms: int | None = Field(default=None, ge=0)
    analysis_prompt_version: str | None = Field(default=None, max_length=16)
    analysis_error_message: str | None = None
    analysis_retry_count: int | None = Field(default=None, ge=0)
    image_analysis: ComplaintImageAnalysis | None = None
    image_analysis_status: AnalysisStatus | None = None
    vision_model: str | None = Field(default=None, max_length=64)
    vision_processing_time_ms: int | None = Field(default=None, ge=0)
    vision_completed_at: datetime | None = None
    vision_started_at: datetime | None = None
    vision_prompt_version: str | None = Field(default=None, max_length=16)
    vision_error_message: str | None = None
    vision_retry_count: int | None = Field(default=None, ge=0)
    is_duplicate: bool | None = None
    duplicate_score: float | None = Field(default=None, ge=0, le=100)
    duplicate_reason: str | None = None
    duplicate_confidence: float | None = Field(default=None, ge=0, le=1)
    matched_complaint_id: str | None = Field(default=None, max_length=128)
    matched_cluster_id: str | None = Field(default=None, max_length=128)
    landmark: str | None = Field(default=None, max_length=256)
    image_base64: str | None = None
    image_mime_type: str | None = Field(default=None, max_length=128)
    audio_base64: str | None = None
    audio_mime_type: str | None = Field(default=None, max_length=128)
    audio_duration_seconds: float | None = Field(default=None, ge=0)
    is_anonymous: bool | None = None
    updated_by: str = Field(default="system", max_length=128)


class ComplaintResponse(ComplaintBase):
    """Complaint document returned from Firestore."""

    id: str
    metadata: DocumentMetadata


class ComplaintSearchFilters(BaseModel):
    """Filter parameters for complaint queries."""

    citizen_email: str | None = None
    constituency: str | None = None
    district: str | None = None
    state: str | None = None
    village_id: str | None = None
    cluster_id: str | None = None
    category: ComplaintCategory | None = None
    category_id: str | None = None
    status: ComplaintStatus | None = None
    priority: ComplaintPriority | None = None
    keyword: str | None = Field(default=None, max_length=64)
    submitted_after: datetime | None = None
    submitted_before: datetime | None = None
